### ISSUE TITLE
Test failed at tc_resolv

### ISSUE THREAD

**Issue by samgooi4189** [alexdalitz/dnsruby#112]
Test failed when running "$ rake test"
```
TestResolv#test_resolv_address_to_name [/home/user/projects/dnsruby/test/tc_resolv.rb:65]:
--- expected
+++ actual
@@ -1 +1 @@
-"google-public-dns-a.google.com"
+"google-public-dns-a.google.com."
```

For ruby built-in resolv,
```
1.9.3-p551 :007 > Resolv.getname("8.8.8.8")
 => "google-public-dns-a.google.com" 
```

Is this a logic fault or test issue?
* Tested on ruby 2.3.3 and ruby 1.9.3

**Comment by alexdalitz** [comment]
Hi -

Thanks for report. I'm traveling just now, but will try to look at this early next week. Tests _should_ pass, and Travis build I think is succeeding, so not sure what problem is. Looks very odd...

**Comment by samgooi4189** [comment]
Thank you for looking into it, cheers.

**Comment by alexdalitz** [comment]
This is very strange.

The issue is that the Name.to_s is returning an absolute name on your system (the trailing “.” at the end of the name). I’m entirely unclear as to why this should be, as the to_s method is very clear that the name is not absolute in the string format.

Can I please ask what version of dnsruby you are testing, and what system you’re running on?

Everything works just fine for Travis, and for master on my OSX system…

Thanks!


Alex.

**Comment by keithrbennett** [comment]
I've now tried several things, and the only reliable variable is whether or not `tc_tcp_pipelining.rb` has been included in the tests.  To save a little time, I ran the tests as:

`    ruby test/ts_online.rb`

To prove the point most quickly, try including in the file list:

- all files except tc_tcp_pipelining.rb (resolv test will pass)
- tc_resolv.rb and tc_tcp_pipelining.rb (resolv test will fail)

Interestingly, the order that the tests ran didn’t matter. This points to the error occurring because of something that happens when the test file is _loaded_ by require_relative. (The way Minitest apparently works is that _all_ test suites are loaded before _any_ of them are run. I guess this makes sense because otherwise it would not be able to randomize the run order.)  So something in the loading of tc_tcp_pipelining appears to be replacing behavior. 

I tried running the following and still encountered the error:

- running the tests on a year old version of the code base
- running the tests on Ubuntu in addition to my usual Mac OS

**Comment by keithrbennett** [comment]
I’ve spent a few more hours investigating this bug.  There are no errors when I run the tests using rake, or in my Linux VM.  The only time I experience errors is on my Mac and when I run the tests using Minitest’s autorun, as in:

`ruby test/ts_online.rb`

I found out that the Dnsruby::Resolv methods were not being called when the error occurs!  Instead, I assume the Ruby runtime’s Resolv class was being called instead.  Since both classes have the same name, and we are using `include Dnsruby`, there is ambiguity as to which one is being referred to – if not according to the technical specifications of the Ruby language, then in the implementation of the Ruby runtime, or possibly native libraries that it calls.

I’ve issued a pull request with the class name changed to DnsrubyResolv, and the problem no longer occurs, although I do get these errors:

  1) Failure:
TestResolv#test_resolv_address_to_name [/Users/kbennett/work/dnsruby/test/tc_resolv.rb:67]:
--- expected
+++ actual
@@ -1 +1 @@
-"google-public-dns-a.google.com"
+"8.8.8.8.in-addr.arpa"



  2) Failure:
TestResolver#test_reverse_lookup [/Users/kbennett/work/dnsruby/test/tc_resolver.rb:141]:
Expected nil to be truthy.

I don’t know whether these errors are the result of the change and do some further research, but if they are due to the calling of the correct class, then they need to be addressed.

I wonder if there’s something about the Mac implementation that’s bad?  I just upgraded to the new Mac Sierra version, 10.12.2.

**Comment by keithrbennett** [comment]
After several hours more of investigating, it now appears that the factors I identified before (Mac OS, tcp_pipelining test) are no longer relevant. I have produced the error on a Linux VM on the cloud, and I have gotten the tests to pass when including tc_pipelining.

There is, however, progress. I learned that when the error occurs, that method is not being called! Instead, the Ruby standard library’s Resolv’s `getname` was being called, even though I could not find a `require 'resolv'` in any of the code.  The following pry excerpt illustrates that 1) the code being called is from Ruby’s resolv, and 2) that it is calling its DefaultResolver to get the name, which explains the error:

```
From: /Users/kbennett/.rvm/rubies/ruby-2.4.0/lib/ruby/2.4.0/resolv.rb @ line 66 Dnsruby::Resolv.getname:

    65: def self.getname(address)
 => 66:   DefaultResolver.getname(address)
    67: end
```

The question is, why and how is Dnsruby::Resolv pointing to ::Resolv (which shouldn’t even be loaded if I understand correctly) instead of Dnsruby::Resolv?

The error seems to be correlated somewhat with the amount of test code that is loaded. At first I thought it was related to the amount of time that had elapsed between the time the tests started and the time of the resolv test, but then I saw that even when the resolv test was the first test run, the error occurred!

By the way, running the tests with the seed below caused tc_resolv to be executed first, saving a lot of time and making it easier to see its result:

`ruby test/ts_online.rb --seed 7819`

**Comment by keithrbennett** [comment]
I've posted a question to StackOverflow about this:

http://stackoverflow.com/questions/42597466/ruby-method-call-made-to-wrong-top-level-module-class

**Comment by keithrbennett** [comment]
I posted this issue to my local Ruby community, and @matt-glover (https://github.com/matt-glover) was kind enough to research the issue.  He noticed that there was an `include Dnsruby` in the top level scope in tc_hs.rb.  When I removed that, the tc_resolv test passed.

I had removed many other such include statements, as referenced in another issue (https://github.com/alexdalitz/dnsruby/issues/113) but missed the one in tc_hs.rb.

I still don't understand why the top level Resolv class dominated and not the Dnsruby::Resolv class though; I would have expected it to be the reverse.

I'm about to issue a pull request for this.
